codalab/worker/download_util.py

```diff
@@ -125,9 +175,118 @@ def _compute_target_info(path, depth):
+            """
+            if not path.startswith("/"):
+                path = "/" + path
+            finfo = tf.getFileInfo(path)
+            if finfo is None:
+                # Not found
+                raise PathException
+            result: TargetInfo = {
+                'name': os.path.basename(path),  # get last part of path
+                'size': finfo.size,
+                'perm': finfo.mode & 0o777,
+                'type': '',
+            }
+            if islink(finfo):
+                result['type'] = 'link'
+                result['link'] = readlink(finfo)
+            elif isfile(finfo):
+                result['type'] = 'file'
+            elif isdir(finfo):
+                result['type'] = 'directory'
+                if depth > 0:
+                    result['contents'] = [
+                        _get_info(path + "/" + file_name, depth - 1)
+                        for file_name in listdir(path)
+                        if file_name != "."
+                    ]
+            return result
+
+        if linked_bundle_path.archive_subpath:
+            # Return the contents of a subpath within a directory.
+            return _get_info(linked_bundle_path.archive_subpath, depth)
+        else:
+            # No subpath, return the entire directory with the bundle
+            # contents in it. The permissions of this directory
+            # cannot be set by the user (the user can only set permissions
+            # of files *within* this directory that are part of the bundle
+            # itself), so we just return a placeholder value of 0o755
+            # for this directory's permissions.
+            file = FileSystems.match([path])[0].metadata_list[0]
+            result: TargetInfo = {
+                'name': linked_bundle_path.bundle_uuid,
+                'type': 'directory',
+                'size': file.size_in_bytes,
+                'perm': 0o755,
+            }
+            if depth > 0:
+                result['contents'] = [
+                    _get_info(file_name, depth - 1)
+                    for file_name in listdir("/")
+                    if file_name != "."
+                ]
+            return result
+
+
+def compute_target_info_beam_descendants_flat(path: str) -> List[TargetInfo]:
+    """Given a path on Azure Blob Storage,
+    returns a flat array of all descendants within that directory in the format
+    [{name, type, size, perm}], where `name` is equal to the full path of each item.
+
+    Also includes an entry for the specified directory with `name` equal to an empty string.
+
+    This function is used by TarSubdirStream in order to determine the list of descendants
+    that exist inside a given subdirectory in a .tar.gz file on Blob Storage.
+    """
+    target_info = _compute_target_info_beam(
+        path, math.inf
+    )  # We want to calculate the target info at infinite depth to expand *all* descendants.
```

Comment:
Couldn't this be very slow?  Might be safer to always have a depth bound (was this the case before?)  For example, what if you have cyclic symlinks?

Reply:
This function (`compute_target_info_beam_descendants_flat`) is used only when streaming a subdirectory within a tar file on Blob Storage to get the initial list of files to stream to create the new .tar file (so normally, when computing target info, there's still a depth bound as usual).

Also, it doesn't resolve symlinks.

The only real slowdown that would happen is if there are too many files within a subdirectory, it would block a bit before starting to stream the file. We could eventually stop this blocking problem by changing this function to return a generator that yields the list of all descendants as well, but perhaps we could do that later once we first see the performance implications to see if it's worth that additional complexity.

Reply:
@percyliang I ended up changing `compute_target_info_blob_descendants_flat` to return a generator instead of a list of all descendants, so listing the contents of a directory for use in `TarSubDirStream` should never block -- I believe this should alleviate your concern.